helix-core/src/main/java/org/apache/helix/cloud/topology/VirtualTopologyGroupStrategy.java

```diff
@@ -0,0 +1,80 @@
+package org.apache.helix.cloud.topology;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import com.google.common.collect.ImmutableMap;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.helix.util.HelixUtil;
+
+public enum VirtualTopologyGroupStrategy {
+
+  /**
+   * A default assignment strategy that is deterministic and stable where:
+   * 1. assignment is guaranteed consistent for same inputs.
+   * 2. number of instance shuffles is reduced.
+   */
+  DEFAULT {
+    @Override
+    public Map<String, Set<String>> computeAssignment(
+        int numGroups, String virtualGroupName, Map<String, Set<String>> zoneMapping) {
+      List<String> sortedInstances = HelixUtil.sortAndFlattenZoneMapping(zoneMapping);
+      Map<String, Set<String>> assignment = new HashMap<>();
+      int instancesPerGroup = sortedInstances.size() / numGroups;
```

Comment:
How would that work? If we have 24 instances over 5 HDF, to assign 7 virtual groups. Notice `ceil(24 / 7) == 4` and `24 / 4 == 6`, 24 instances are effectively only assigned to 6 groups.

Reply:
Not exactly, the reason you get the conclusion of 6 groups is that you sort the instance and compute groupIndex with "instanceInd / instancesPerGroup". My original idea is that we can directly use mod to get the index.
However, according to what we have discussed yesterday, it might be problematic to assign instances of one physical zone to multiple virtual groups. It may effectively downgrade fault tolerance. So I assume you will have to change the design here. Let's discuss further when your new design is ready : )